Ingesting Historical Data

MichaelMedford opened the pull request:
This PR adds support for ingesting historical one-minute data into InfluxDB the first time that Vuegraf is run. A new parameter is added to the configuration, `historyDays`, to support this feature. Emporia currently retains one-minute data for 7 days, and therefore `historyDays` must be less than or equal to `7`. If `historyDays` is set to `0`, no historical data will be ingested into InfluxDB.

If this PR is accepted, I would be happy to create another PR for ingestion of historical one-hour data. Emporia currently retains one-hour data for the lifetime of the device.

jertel commented on src/vuegraf/vuegraf.py:
```diff
@@ -138,6 +138,29 @@ def extractDataPoints(device, usageDataPoints):
+            if kwhUsage is not None:
+                timestamp = startTime + datetime.timedelta(minutes=index)
+                watts = float(minutesInAnHour * wattsInAKw) * kwhUsage
+                usageDataPoints.append(createDataPoint(account, chanName, watts, timestamp, True))
```
I noticed the `True` arg is being passed into createDataPoint() here. That indicates that this is _detailed_ data, which is a way for users to distinguish minute data vs second (detailed) data. Since this history data is per-minute, it might make more sense to pass `False` here instead of `True`.

MichaelMedford replied:
Thanks for catching that!

jertel commented on src/vuegraf/vuegraf.py:
```diff
@@ -138,6 +138,29 @@ def extractDataPoints(device, usageDataPoints):
                     usageDataPoints.append(createDataPoint(account, chanName, watts, timestamp, True))
                     index += 1
 
+def extractHistoricalDataPoints(device, startTime, stopTime, usageDataPoints):
```
There's a lot of duplicated logic with the above `extractDataPoints` function. Did you consider adjust the other function to accept start/stop params, and perhaps a historic flag? The primary concern with the duplicated code is that a future PR might not realize there are two functions that perform similar duties and so one could get updated, but not the other.

MichaelMedford replied:
You're definitely right that it would be better for all of the logic to be contained within a single function. I chose to use `extractDataPoints ` for historical data whenever the `startTime` and `endTime` flags are specified, instead of a `history` flag. I prefer this because it is less redundant than creating a new variable only for the purpose of denoting `history` "mode", when the other two variables are only set for this purpose.

MichaelMedford commented:
@jertel After addressing your comments, I have added in some additional code to (1) pause between day calls to the API to avoid errors from Emporia and (2) correctly catch keyboard interrupts within the loop.

jertel commented on src/vuegraf/vuegraf.py:
```diff
@@ -110,7 +110,7 @@ def createDataPoint(account, chanName, watts, timestamp, detailed):
     return dataPoint
 
-def extractDataPoints(device, usageDataPoints):
+def extractDataPoints(device, usageDataPoints, startTime=None, stopTime=None):
```
This is looking pretty good. I think if you change these parameter names to `historyStartTime` and `historyStopTime` it will avoid confusion.

MichaelMedford commented on src/vuegraf/vuegraf.py:
```diff
@@ -119,24 +119,41 @@ def extractDataPoints(device, usageDataPoints):
         if kwhUsage is not None:
-            chanName = lookupChannelName(account, chan)
             watts = float(minutesInAnHour * wattsInAKw) * kwhUsage
             timestamp = stopTime
```
@jertel This `stopTime` is not explicitly passed into the function. That could be implemented on a future PR.

jertel replied:
Yes there are multiple cases of this. It started out as a few lines of python scripting but feature requests have caused it to become more complex. A refactor is probably due at some point.

jertel commented on src/vuegraf/vuegraf.py:
```diff
@@ -233,6 +253,21 @@ def extractDataPoints(device, usageDataPoints):
+                    if history:
+                        for day in range(historyDays):
+                            info('Loading historical data: {} day ago'.format(day+1))
+                            histroyStartTime = stopTime - datetime.timedelta(seconds=3600*24*(day+1))
```
`histroyStartTime` typo

jertel reviewed (approved):
Thanks for taking the time to add this feature!
